internal/querycoordv2/utils/target.go

```diff
@@ -0,0 +1,78 @@
+package utils
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+	"github.com/milvus-io/milvus/internal/querycoordv2/meta"
+)
+
+func IsNextTargetReadyForCollection(targetMgr *meta.TargetManager, distMgr *meta.DistributionManager, metaInfo *meta.Meta, collectionID int64) bool {
+	replicaNum := len(metaInfo.ReplicaManager.GetByCollection(collectionID))
+
+	// check channel first
+	channelNames := targetMgr.GetDmChannelsByCollection(collectionID, meta.NextTarget)
+	if len(channelNames) == 0 {
+		// next target is empty, no need to update
+		return false
+	}
+	for name := range channelNames {
+		if replicaNum > len(distMgr.GetChannelDist(name)) {
+			return false
+		}
+	}
+	// then check streaming segment
+	streamingSegments := targetMgr.GetStreamingSegmentsByCollection(collectionID, meta.NextTarget)
+	for ID := range streamingSegments {
+		if replicaNum > len(distMgr.GetGrowingSegmentDist(ID)) {
+			return false
+		}
+	}
+
+	// and last check historical segment
+	historicalSegments := targetMgr.GetHistoricalSegmentsByCollection(collectionID, meta.NextTarget)
+	for ID := range historicalSegments {
+		if replicaNum > len(distMgr.GetSealedSegmentDist(ID)) {
```

Comment:
Checking the loaded segment count is not enough to judge this segment is loaded in different replica. Because we try to load a segment first when doing compaction. So in a replica, it is possible that a segment exists in two nodes at the same time.

Reply:
BTW, we do not support moving shard leader to another node automatically or manually for now. But after supporting it, checking the channel count like above is not enough neither.

Reply:
agree, should use same logic with `CollectionObserver` to check load status.